gnucash/gnome-utils/gnc-account-sel.c

```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
+    menu = gtk_menu_new ();
+    h_placeholder = gtk_check_menu_item_new_with_mnemonic (_("Hide _Placeholder Accounts"));
+    gtk_check_menu_item_set_active (GTK_CHECK_MENU_ITEM(h_placeholder), gas->hide_placeholder);
+    h_hidden = gtk_check_menu_item_new_with_mnemonic (_("Hide _Hidden Accounts"));
+    gtk_check_menu_item_set_active (GTK_CHECK_MENU_ITEM(h_hidden), gas->hide_hidden);
+    gtk_menu_shell_append (GTK_MENU_SHELL(menu), h_placeholder);
+    gtk_menu_shell_append (GTK_MENU_SHELL(menu), h_hidden);
+    gtk_widget_show_all (menu);
+
+    g_signal_connect (G_OBJECT(h_placeholder), "toggled",
+                      G_CALLBACK(toggle_placeholder_cb), gas);
+    g_signal_connect (G_OBJECT(h_hidden), "toggled",
+                      G_CALLBACK(toggle_hidden_cb), gas);
+
+    gtk_menu_popup_at_pointer (GTK_MENU(menu), (GdkEvent *)event);
+}
 
-    atnd.gas        = gas;
-    atnd.outList    = NULL;
+static gboolean
+account_is_included (GNCAccountSel *gas, Account *acc)
+{
+    gboolean included = TRUE;
 
-    g_list_foreach (accts, gas_filter_accounts, (gpointer)&atnd);
-    g_list_free (accts);
-    atnd.outList = g_list_reverse (atnd.outList);
+    if (gas->acctExcludeList)
+    {
+        if (g_list_find (gas->acctExcludeList, acc) != NULL)
+            included = FALSE;
+    }
 
-    gtk_list_store_clear (gas->store);
-    for (ptr = atnd.outList, i = 0; ptr; ptr = g_list_next(ptr), i++)
+    /* Filter as we've been configured to do. */
+    if (gas->acctTypeFilters)
     {
-        acc = ptr->data;
-        name = gnc_account_get_full_name (acc);
-        gtk_list_store_append (gas->store, &iter);
-        gtk_list_store_set (gas->store, &iter,
-                            ACCT_COL_NAME, name,
-                            ACCT_COL_PTR,  acc,
-                            -1);
-        if (g_utf8_collate (name, currentSel) == 0)
-            active = i;
-        g_free (name);
+        /* g_list_find is the poor-mans '(member ...)', especially
+         * easy when the data pointers in the list are just casted
+         * account type identifiers. */
+        if (g_list_find (gas->acctTypeFilters,
+                         GINT_TO_POINTER(xaccAccountGetType (acc))) == NULL)
+        {
+            included = FALSE;
+        }
     }
 
-    /* If the account which was in the text box before still exists, then
-     * reset to it. */
-    if (active != -1)
-        gtk_combo_box_set_active (GTK_COMBO_BOX(gas->combo), active);
+    if (gas->acctCommodityFilters)
+    {
+        if (g_list_find_custom (gas->acctCommodityFilters,
+                                GINT_TO_POINTER(xaccAccountGetCommodity (acc)),
+                                gnc_commodity_compare_void) == NULL)
+        {
+            included = FALSE;
+        }
+    }
+    return included;
+}
 
-    g_signal_handlers_unblock_by_func (gas->combo, combo_changed_cb , gas);
+static gboolean
+visible_func (GtkTreeModel *model, GtkTreeIter *iter, gpointer user_data)
```

Comment:
Let's name this something a little more expressive like `account_is_visible_cb`.